### ISSUE TITLE
Setting secure SSLContext::DEFAULT_PARAMS options isn't working

### ISSUE THREAD

**Issue by nevans** [hashicorp/vault-ruby#250]
This isn't doing what it seems to think it's doing:
https://github.com/hashicorp/vault-ruby/blob/0a0a309602e017fcfca70e4eb7d3919f9b48f903/lib/vault.rb#L20-L26 

Specifically, `opts` as an `Integer` (`Fixnum` in pre-2.4 ruby) is a _block local variable_ inside that `tap` block. It is never assigned to `OpenSSL::SSL::SSLContext::DEFAULT_PARAMS`. All this does is change the local variable!  You can test this, e.g. in the console, like so:

```ruby
OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options] = OpenSSL::SSL::OP_DONT_INSERT_EMPTY_FRAGMENTS
# => 2048
OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options].tap do |opts| 
   opts &= ~OpenSSL::SSL::OP_DONT_INSERT_EMPTY_FRAGMENTS if defined?(OpenSSL::SSL::OP_DONT_INSERT_EMPTY_FRAGMENTS)    
   opts |= OpenSSL::SSL::OP_NO_COMPRESSION if defined?(OpenSSL::SSL::OP_NO_COMPRESSION)    
   opts |= OpenSSL::SSL::OP_NO_SSLv2 if defined?(OpenSSL::SSL::OP_NO_SSLv2)    
   opts |= OpenSSL::SSL::OP_NO_SSLv3 if defined?(OpenSSL::SSL::OP_NO_SSLv3)    
end   
# => 2048
OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options]
# => 2048
```

What we want to do instead looks more like the following, which updates the hash inside the block:

```ruby
    # Set secure SSL options
    OpenSSL::SSL::SSLContext::DEFAULT_PARAMS.tap do |defaults|
      defaults[:options] &= ~OpenSSL::SSL::OP_DONT_INSERT_EMPTY_FRAGMENTS if defined?(OpenSSL::SSL::OP_DONT_INSERT_EMPTY_FRAGMENTS)
      defaults[:options] |= OpenSSL::SSL::OP_NO_COMPRESSION if defined?(OpenSSL::SSL::OP_NO_COMPRESSION)
      defaults[:options] |= OpenSSL::SSL::OP_NO_SSLv2 if defined?(OpenSSL::SSL::OP_NO_SSLv2)
      defaults[:options] |= OpenSSL::SSL::OP_NO_SSLv3 if defined?(OpenSSL::SSL::OP_NO_SSLv3)
    end
```

or the following, which calls the block, then assigns the result (this is roughly stdlib openssl does, which is where I assume this code was copied from):
```ruby
    # Set secure SSL options
    OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options] = -> {
      opts = OpenSSL::SSL::SSLContext::DEFAULT_PARAMS[:options]
      opts &= ~OpenSSL::SSL::OP_DONT_INSERT_EMPTY_FRAGMENTS if defined?(OpenSSL::SSL::OP_DONT_INSERT_EMPTY_FRAGMENTS)
      opts |= OpenSSL::SSL::OP_NO_COMPRESSION if defined?(OpenSSL::SSL::OP_NO_COMPRESSION)
      opts |= OpenSSL::SSL::OP_NO_SSLv2 if defined?(OpenSSL::SSL::OP_NO_SSLv2)
      opts |= OpenSSL::SSL::OP_NO_SSLv3 if defined?(OpenSSL::SSL::OP_NO_SSLv3)
    }.call
```

On the other hand, these are already the defaults, since:
* (I think?) OpenSSL 1.1.0 (Aug 2016, although unfortunately some people may be able to upgrade ruby more rapidly than openssl)
* ruby since Oct 2014 ([Changing defaults announcement](https://www.ruby-lang.org/en/news/2014/10/27/changing-default-settings-of-ext-openssl/))
  * [the bugreport that led to changing the defaults and backporting](https://bugs.ruby-lang.org/issues/9424)
  * ruby v2.2.0 ([trunk rev 45274](https://github.com/ruby/ruby/commit/699b209cf8cf11809620e12985ad33ae33b119ee)), Mar 2014
  * [ruby v2.1.4](https://www.ruby-lang.org/en/news/2014/10/27/ruby-2-1-4-released/), Oct 2014
  * [ruby v2.0.0-p594](https://www.ruby-lang.org/en/news/2014/10/27/ruby-2-0-0-p594-is-released/), Oct 2014
  * [ruby v1.9.3-p550](https://www.ruby-lang.org/en/news/2014/10/27/ruby-1-9-3-p550-is-released/), Oct 2014

So hopefully most users simply inherited the good defaults anyway!

**Comment by Valarissa** [comment]
Taking a look now, sorry for the lack of response.

**Comment by Valarissa** [comment]
Thanks~! This was totally the case! It should be addressed in #255

**Comment by nevans** [comment]
I was looking up something else and noticed this issue was still open.  It was definitely fixed by #255!  :)
